### PR TITLE
Add a pruner based on ASHA

### PR DESCRIPTION
**Pull request by sile**
This PR contains the following changes (mainly):
- Created `optuna/pruners/` directory and split `pruners.py` into `pruners/base.py` and `pruners/median.py`
- Added `SuccessiveHalvingPruner` class
- Added test cases for `SuccessiveHalvingPruner`

### CODE REVIEW

File: optuna/pruners/asha.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+import math
+
+from optuna.pruners.base import BasePruner
+from optuna.storages import BaseStorage  # NOQA
+from optuna.structs import FrozenTrial  # NOQA
+from typing import List  # NOQA
+
+
+class ASHAPruner(BasePruner):
```
Comment (iwiwi) [inline optuna/pruners/asha.py]: Do you prefer `ASHAPruner` over other descriptive names such as `SuccessiveHalvingPruner`? I think the name `ASHA` is not very popular yet.

File: optuna/pruners/asha.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            the same process with a new ``rung``.
+        eta:
+            A parameter for specifying reduction factor of promotable trials.
+            At the completion point of the each rung, about ``1/eta`` trials will be pruned.
+        s:
+            A parameter for specifying the completion point of the first rung.
+            The point is determined by the expression ``r * (eta ** s)``.
+    """
+
+    def __init__(self, r=1, eta=4, s=0):
```
[the remark this answers is not in the document]
Comment (iwiwi) [reply optuna/pruners/asha.py]: I prefer option (b), as it would make both library and user codes more readable. I appreciate if you could describe the corresponding variable names in the paper in the docstring.

File: optuna/pruners/asha.py (hunk 1 above)
Comment (sile) [reply optuna/pruners/asha.py]: Done: https://github.com/pfnet/optuna/pull/236/commits/e184402baaf4ad41cad6983052032188e13f583b

**Comment by sile** [comment]
@g-votte @toshihikoyanase @iwiwi 

I updated documentation of `SuccessiveHalvingPruner`. So it is ready to be reviewed.

**Comment by sile** [comment]
@g-votte @toshihikoyanase
I updated documents.  Please tell me if there are some points to improve.

File: optuna/pruners/successive_halving.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            # the following trials even if the rung contains only trials less than `eta`.
+            #
+            # So, we allow the first `eta - 1` trials to become promotable
+            # if its value is the smallest among the trials that already have reached the rung.
+            promotable_idx = 0
+
+        # TODO(ohta): Deal with maximize direction.
+        return value <= competing_values[promotable_idx]
+
+    def _get_current_rung(self, trial):
```
Comment (toshihikoyanase) [inline optuna/pruners/successive_halving.py]: This method can be static because it does not use any member variables/methods.

File: optuna/pruners/successive_halving.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            if all_trials is None:
+                all_trials = storage.get_all_trials(study_id)
+
+            storage.set_trial_system_attr(trial_id, completed_rung_key(rung), value)
+            if not self._is_promotable(storage, rung, value, all_trials):
+                return True
+
+            rung += 1
+
+    def _is_promotable(self, storage, rung, value, all_trials):
```
Comment (toshihikoyanase) [inline optuna/pruners/successive_halving.py]: `storage` is not used in this method.

**Comment by sile** [comment]
@g-votte @toshihikoyanase 
Thank you for your careful review. I updated documentation and some functions to reflect your comments.

File: optuna/pruners/successive_halving.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            >>>     ...
+            >>>
+            >>> study = create_study(pruner=SuccessiveHalvingPruner())
+            >>> study.optimize(objective)
+
+    Args:
+        min_resource:
+            A parameter for specifying the minimum resource allocated to a trial
+            (in the `paper <http://arxiv.org/abs/1810.05934>`_ this parameter is
+            referred to as "r").
```
Comment (toshihikoyanase) [inline optuna/pruners/successive_halving.py]: Because back quotes are used to escape numerical expressions like `R` in L19, the quotation marks around `r` should be replaced with backquotes. Please take care of `eta` and `s` in addition to `r`.

Comment (sile) [reply optuna/pruners/successive_halving.py]: Thank you for pointing it out. I fixed this at the commit https://github.com/pfnet/optuna/pull/236/commits/429e27e392594ad3344d955f98d6adf69d341974.

File: tests/pruners_tests/test_successive_halving.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    assert not pruner.prune(study.storage, study.study_id, trial.trial_id, step=1)
+    assert 'completed_rung_0' not in trial.system_attrs
+
+    trial.report(1, step=2)
+    assert not pruner.prune(study.storage, study.study_id, trial.trial_id, step=2)
+    assert 'completed_rung_0' in trial.system_attrs
+    assert 'completed_rung_1' not in trial.system_attrs
+
+
+def test_successive_halving_pruner_reduction_factor_parameter():
```
Comment (toshihikoyanase) [inline tests/pruners_tests/test_successive_halving.py]: Please add a type hint.

File: tests/pruners_tests/test_successive_halving.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    assert not pruner.prune(study.storage, study.study_id, trial.trial_id, step=2)
+    assert 'completed_rung_1' not in trial.system_attrs
+
+    trial.report(1, step=3)
+    assert not pruner.prune(study.storage, study.study_id, trial.trial_id, step=3)
+    assert 'completed_rung_1' in trial.system_attrs
+    assert 'completed_rung_2' not in trial.system_attrs
+
+
+def test_successive_halving_pruner_min_early_stopping_rate_parameter():
```
Comment (toshihikoyanase) [inline tests/pruners_tests/test_successive_halving.py]: ditto.

File: tests/pruners_tests/test_successive_halving.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    assert 'completed_rung_3' in trial.system_attrs
+    assert 'completed_rung_4' not in trial.system_attrs
+
+
+def test_successive_halving_pruner_min_resource_parameter():
+    # type: () -> None
+
+    study = optuna.study.create_study()
+
+    # min_resource=1: The rung 0 ends at step 1.
```
Comment (toshihikoyanase) [inline tests/pruners_tests/test_successive_halving.py]: I think we can add an edge case where `min_resource < 1` to confirm `SuccessiveHalvingPruner.__init__()` raises an appropriate exception.

File: tests/pruners_tests/test_successive_halving.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    # type: () -> None
+
+    study = optuna.study.create_study()
+    trial = optuna.trial.Trial(study, study.storage.create_new_trial_id(study.study_id))
+    trial.report(1, 1)
+    pruner = optuna.pruners.SuccessiveHalvingPruner(
+        min_resource=1, reduction_factor=2, min_early_stopping_rate=0)
+
+    # A pruner is not activated at a first trial.
+    assert not pruner.prune(study.storage, study.study_id, trial.trial_id, step=1)
```
Comment (toshihikoyanase) [inline tests/pruners_tests/test_successive_halving.py]: I could not find the case where the target value of the pruner is `NaN`. Please add a test function like `test_successive_halving_pruner_with_nan` to confirm trials with `NaN` values are pruned without any exceptions.

File: tests/pruners_tests/test_successive_halving.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+    # Report a trial that has the lowest value.
+    trial = optuna.trial.Trial(study, study.storage.create_new_trial_id(study.study_id))
+    trial.report(0.05, step=7)
+    pruner.prune(study.storage, study.study_id, trial.trial_id, step=7)
+    assert 'completed_rung_2' in trial.system_attrs
+    assert 'completed_rung_3' not in trial.system_attrs
+
+
+def test_successive_halving_pruner_first_trial_always_wins():
```
Comment (g-votte) [inline tests/pruners_tests/test_successive_halving.py]: Because of "wins", I think this method name is a bit misleading. As there's no competition among trials in this test scenario, how about `test_successive_halving_pruner_first_trial_is_not_pruned`?

File: tests/pruners_tests/test_successive_halving.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    trial = optuna.trial.Trial(study, study.storage.create_new_trial_id(study.study_id))
+    # A pruner is not activated if a trial has no intermediate values.
+    assert not pruner.prune(study.storage, study.study_id, trial.trial_id, step=1)
+
+    trial.report(2, 1)
+    # A pruner is activated if a trial has an intermediate value.
+    assert pruner.prune(study.storage, study.study_id, trial.trial_id, step=1)
+
+
+def test_successive_halving_pruner_up_to_third_rung():
```
Comment (g-votte) [inline tests/pruners_tests/test_successive_halving.py]: With this method name, I feel as if `third` were a boundary value. How about renaming this method to `test_successive_halving_pruner_rung_check`?

**Comment by sile** [comment]
@toshihikoyanase @g-votte I updated `test_successive_halving.py` for reflecting your review comments.
